docs/token-contracts/fa12/4-fa12-archetype.md

```diff
@@ -0,0 +1,447 @@
+
+```
+$ archetype --set-caller-init tz1dEezCSqMVsmfnaAZS2rBJZfdFCZEyA7DP -t michelson-storage fa12.arl
+(Pair {  } { Elt "tz1dEezCSqMVsmfnaAZS2rBJZfdFCZEyA7DP" 10000000 })
+```
+
+The address passed as the initial caller (here `tz1dEezCSqMVsmfnaAZS2rBJZfdFCZEyA7DP`) is the address that possesses the total number of tokens (here 100000000). The generated initial storage is used to generate the origination command.
+
+The following command originates the contract:
+```
+$ tezos-client -S -A testnet-tezos.giganode.io -P 443 originate contract fa12 transferring 0 from alice running fa12.tz --init '(Pair {  } { Elt "tz1dEezCSqMVsmfnaAZS2rBJZfdFCZEyA7DP" 10000000 })' --burn-cap 0 --force -D
+```
+
+The cost of origination (at the time of writing) is **2.5ꜩ** on the testnet.
+
+It uses the [Giganode](https://tezos.giganode.io/) Tezos nodes infrastructure.
+
+> Install [Tezos client](../../../docs/setup/1-tezos-client).
+
+## Formal properties
+This section presents the properties verified by the originated contract.
+
+The contract declares two [assets](https://docs.archetype-lang.org/archetype-language/data-model):
+
+```archetype
+asset ledger identified by holder to big_map {
+  holder     : address;
+  tokens     : nat = 0;
+} initialized by {
+  { holder = caller; tokens = totalsupply }
+}
+```
+The `ledger` asset is the cap table: it holds the number of tokens for each token holder.
+`totalsupply` is the initial number of tokens held by the originator of the contract.
+
+```archetype
+asset allowance identified by addr_owner addr_spender to big_map {
+  addr_owner       : address;
+  addr_spender     : address;
+  amount           : nat;
+}
+```
+The `allowance` asset stores the amount of tokens that can be spent by `addr_spender` on the behalf of `addr_owner`.
+
+### Contract invariant
+
+```archetype
+ledger.sum(tokens) = totalsupply
+```
+> The total number of tokens is equal to `totalsupply` regardless of the state of the contract storage.
+
+### transfer
+
+```archetype
+entry %transfer (%from : address, %to : address, value : nat) {
+  require {
+    r1 otherwise "NotEnoughBalance" : ledger[%from].tokens >= value;
+  }
+  effect {
+    if caller <> %from then (
+      var current = allowance[(%from, caller)].amount;
+      dofailif(current < value, ("NotEnoughAllowance", ((value, current))));
+      allowance.update((%from, caller), { amount -=  value });
+    );
+    ledger.update(%from, { tokens -= value });
+    ledger.addupdate(%to, { tokens += value });
+  }
+}
+```
+#### 1/ Effect on `ledger`
+
+##### 1.1/ Effect on `%from` token holder
+```archetype
+%from <> %to ->
+let some before_ledger_from = before.ledger[%from] in
+let some after_ledger_from  = ledger[%from] in
+  after_ledger_from = { before_ledger_from with
+    tokens = (before_ledger_from.tokens - value)
+  }
+otherwise false otherwise false
+```
+> When the `%to` address is different from the `%from` address, the number of tokens `%to` possesses is decread by `value`.
+
```

Comment:
The first one could have a bit more hand-holding: `->` is implication, what is `before`, why are there 2 `otherwise`s?, what do they really mean? etc.

Reply:
Done; added the paragraph:
```
Elements of formal property language:
* `... -> ...` is the logical implication; it reads 'if ... then ...' (or '... implies ...')
* `let some ... in ... otherwise ...` is necessary because the `[  ]` operator on asset collection returns an option of asset; the `otherwise` branch is used to state property when the asset does not exist
* `before` refers to the state of the asset collection *before* execution of the entry point; without modifier, the default asset collection identifier refers to the state of the collection *after* execution

Please refer to the [documentation](https://docs.archetype-lang.org/archetype-language/contract-specification) for further explanation.

Thus the property reads: if `%from` is different from `%to` then the `%from` token holder *after* execution is equal to the `%from` token holder *before* execution except for the `tokens` field which is decreased by `value`.

Postconditions apply *when the entry point does not fail*.

The `transfer` entry fails when the `%from` token holder is not found in `ledger`. Therefore the contradiction property `false` is stated when the `%from` token holder is not found, whether *before* or *after* execution.

```